New version (5.1) breaks older installations

### Prerequisites

* [ x ] Can you reproduce?
* [ x] Are you running the latest version of SoftEtherVPN?

**SoftEther version:**
**Component:** Server
**Operating system:** Ubuntu 16.04
**Architecture:** 64 bit

[In case it's a computer with known specs, such as the Raspberry Pi, you can specify it omitting the details.]
**Processor:** Intel Pentium D (not that it matters to this issue at all)

### Description

Update broke current install.

**Expected behavior:**
Installed correctly and no error

**Actual behavior:**
The installation failed with the following:

 ```
Jul 21 12:52:10 ubuntu systemd[1]: Starting LSB: SoftEther VPN service...
 -- Subject: Unit softether-vpnserver.service has begun start-up
 -- Defined-By: systemd
 -- Support: http://lists.freedesktop.org/mailman/listinfo/systemd-devel
 --
 -- Unit softether-vpnserver.service has begun starting up.
 Jul 21 12:52:10 ubuntu softether-vpnserver[16510]: Starting SoftEtherVPN Server: vpnserver/usr/bin/vpnserver: 2: /usr/bin/vpnserver: /opt/vpnserver/vpnserver: not found
 Jul 21 12:52:10 ubuntu systemd[1]: softether-vpnserver.service: Control process exited, code=exited status=2
 Jul 21 12:52:10 ubuntu systemd[1]: Failed to start LSB: SoftEther VPN service.
 -- Subject: Unit softether-vpnserver.service has failed
 -- Defined-By: systemd
 -- Support: http://lists.freedesktop.org/mailman/listinfo/systemd-devel
 --
 -- Unit softether-vpnserver.service has failed.
 --
 -- The result is failed. 
```

### Steps to reproduce

1. Install version 5.1 over 4.23
2. Watch the install and see it fail.

I managed to recover my installation by a backup of the config file and grabbing a copy of an older version deb in the repository of paskal.

`make[1]: Entering directory '/<<PKGBUILDDIR>>'
echo '#!/bin/sh'                              > debian/usr/bin/vpnserver
/bin/sh: 1: cannot create debian/usr/bin/vpnserver: Directory nonexistent
debian/rules:20: recipe for target 'override_dh_auto_install' failed
make[1]: *** [override_dh_auto_install] Error 2
make[1]: Leaving directory '/<<PKGBUILDDIR>>'
debian/rules:14: recipe for target 'binary' failed
make: *** [binary] Error 2
dpkg-buildpackage: error: fakeroot debian/rules binary subprocess returned exit status 2`

1. Пока не понятно как установить нужную версию 4.23.9647 из ppa, быстрый гугл не помог и скачать не нашел где можно, весь ppa перерыл.

ubuntu 18.04:
2. при сборке из исходников скаченных с офф сайт выдает ошибку:
collect2: error: ld returned 1 exit status
Makefile:18: recipe for target 'i_read_and_agree_the_license_agreement' failed
make[1]: *** [i_read_and_agree_the_license_agreement] Error 1
3. при сборке из гитхаба https://github.com/SoftEtherVPN/SoftEtherVPN/ собирается, но при подключении к серверу пишет: Protocol Error is occurring.

В Ubuntu 14.04 из официальных исходников ставится и работает 4.25 RTM.

"при сборке из исходников скаченных с офф сайт выдает ошибку:
collect2: error: ld returned 1 exit status
Makefile:18: recipe for target 'i_read_and_agree_the_license_agreement' failed
make[1]: *** [i_read_and_agree_the_license_agreement] Error " - можете по шагам описать, как вы собираете ?

по крайней мере на гитлабе билды зеленые, все собирается.
локально у меня тоже

--- a/.travis.yml
+++ b/.travis.yml
@@ -31,6 +31,7 @@ matrix:
         - ./configure
         - make -C tmp
         - otool -L build/vpnserver/vpnserver
+        - sudo make -C tmp install
 
 addons:
   apt:
@@ -55,4 +56,5 @@ script:
   - ./configure
   - make -C tmp
   - ldd build/vpnserver/vpnserver
+  - sudo LD_LIBRARY_PATH="${HOME}/opt/lib:${LD_LIBRARY_PATH:-}" make -C tmp install
   - if [ "${BUILD_DEB}" = "1" ]; then make package -C tmp; fi

--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -1,18 +1,38 @@
 include(GNUInstallDirs)
 
-macro(install_wrapper_script component target destination)
-  get_filename_component(filename ${target} NAME)
+# Creates wrapper scripts and installs them in the user's binaries directory, which is usually "/usr/local/bin".
+# This is required because symlinks use the folder they are in as working directory.
+macro(install_wrapper_script component target)
+  get_filename_component(file_name ${target} NAME)
 
-  file(WRITE ${CMAKE_SOURCE_DIR}/tmp/script/${filename} "#!/bin/sh\n")
-  file(APPEND ${CMAKE_SOURCE_DIR}/tmp/script/${filename} "${target} \"$@\"\n")
-  file(APPEND ${CMAKE_SOURCE_DIR}/tmp/script/${filename} "exit $?\n")
+  file(WRITE ${CMAKE_SOURCE_DIR}/tmp/script/${file_name} "#!/bin/sh\n")
+  file(APPEND ${CMAKE_SOURCE_DIR}/tmp/script/${file_name} "${target} \"$@\"\n")
+  file(APPEND ${CMAKE_SOURCE_DIR}/tmp/script/${file_name} "exit $?\n")
 
-  install(FILES ${CMAKE_SOURCE_DIR}/tmp/script/${filename}
+  install(FILES ${CMAKE_SOURCE_DIR}/tmp/script/${file_name}
     COMPONENT ${component}
-    DESTINATION ${destination}
+    DESTINATION ${CMAKE_INSTALL_FULL_BINDIR}
     PERMISSIONS OWNER_READ OWNER_WRITE OWNER_EXECUTE GROUP_READ GROUP_EXECUTE WORLD_READ WORLD_EXECUTE
   )
 endmacro(install_wrapper_script)
+
+macro(install_systemd_service component file_path binary_path)
+    get_filename_component(file_name ${file_path} NAME)
+    get_filename_component(binary_directory ${binary_path} DIRECTORY)
+
+    file(READ ${file_path} FILE_CONTENT)
+    string(REPLACE "[DIRECTORY]" ${binary_directory} FILE_CONTENT ${FILE_CONTENT})
+    string(REPLACE "[BINARY]" ${binary_path} FILE_CONTENT ${FILE_CONTENT})
+    file(WRITE ${CMAKE_SOURCE_DIR}/tmp/systemd/${file_name} ${FILE_CONTENT})
+
+  if(EXISTS "/lib/systemd/system")
+    install(FILES ${CMAKE_SOURCE_DIR}/tmp/systemd/${file_name}
+      COMPONENT ${component}
+      DESTINATION "/lib/systemd/system"
+      PERMISSIONS OWNER_READ OWNER_WRITE GROUP_READ WORLD_READ
+    )
+  endif()
+endmacro(install_systemd_service)
 
 if(CMAKE_BUILD_TYPE STREQUAL "Debug")
   add_definitions(-D_DEBUG -DDEBUG)
@@ -128,7 +148,7 @@ add_custom_command(TARGET hamcore-archive-build
   COMMAND ${CMAKE_COMMAND} -E copy ${CMAKE_SOURCE_DIR}/tmp/hamcore.se2 ${VPNCMD_RUNTIME_OUTPUT_DIRECTORY}
 )
 
-# Copy "vpnserver" directory to /usr/lib(exec)/softether/
+# Copy "vpnserver" directory to /usr/lib(exec)/softether/, install launch script and systemd service
 install(DIRECTORY ${VPNSERVER_RUNTIME_OUTPUT_DIRECTORY}
   COMPONENT "vpnserver"
   DESTINATION "${CMAKE_INSTALL_LIBEXECDIR}/softether"
@@ -136,7 +156,10 @@ install(DIRECTORY ${VPNSERVER_RUNTIME_OUTPUT_DIRECTORY}
   PERMISSIONS OWNER_READ OWNER_WRITE OWNER_EXECUTE GROUP_READ GROUP_EXECUTE WORLD_READ WORLD_EXECUTE
 )
 
-# Copy "vpnclient" directory to /usr/lib(exec)/softether/
+install_wrapper_script("vpnserver" "${CMAKE_INSTALL_FULL_LIBEXECDIR}/softether/vpnserver/vpnserver")
+install_systemd_service("vpnserver" "${CMAKE_SOURCE_DIR}/systemd/softether-vpnserver.service" "${CMAKE_INSTALL_FULL_LIBEXECDIR}/softether/vpnserver/vpnserver")
+
+# Copy "vpnclient" directory to /usr/lib(exec)/softether/, install launch script and systemd service
 install(DIRECTORY ${VPNCLIENT_RUNTIME_OUTPUT_DIRECTORY}
   COMPONENT "vpnclient"
   DESTINATION "${CMAKE_INSTALL_LIBEXECDIR}/softether"
@@ -144,7 +167,10 @@ install(DIRECTORY ${VPNCLIENT_RUNTIME_OUTPUT_DIRECTORY}
   PERMISSIONS OWNER_READ OWNER_WRITE OWNER_EXECUTE GROUP_READ GROUP_EXECUTE WORLD_READ WORLD_EXECUTE
 )
 
-# Copy "vpnbridge" directory to /usr/lib(exec)/softether/
+install_wrapper_script("vpnclient" "${CMAKE_INSTALL_FULL_LIBEXECDIR}/softether/vpnclient/vpnclient")
+install_systemd_service("vpnclient" "${CMAKE_SOURCE_DIR}/systemd/softether-vpnclient.service" "${CMAKE_INSTALL_FULL_LIBEXECDIR}/softether/vpnclient/vpnclient")
+
+# Copy "vpnbridge" directory to /usr/lib(exec)/softether/, install launch script and systemd service
 install(DIRECTORY ${VPNBRIDGE_RUNTIME_OUTPUT_DIRECTORY}
   COMPONENT "vpnbridge"
   DESTINATION "${CMAKE_INSTALL_LIBEXECDIR}/softether"
@@ -152,7 +178,10 @@ install(DIRECTORY ${VPNBRIDGE_RUNTIME_OUTPUT_DIRECTORY}
   PERMISSIONS OWNER_READ OWNER_WRITE OWNER_EXECUTE GROUP_READ GROUP_EXECUTE WORLD_READ WORLD_EXECUTE
 )
 
-# Copy "vpncmd" directory to /usr/lib(exec)/softether/
+install_wrapper_script("vpnbridge" "${CMAKE_INSTALL_FULL_LIBEXECDIR}/softether/vpnbridge/vpnbridge")
+install_systemd_service("vpnbridge" "${CMAKE_SOURCE_DIR}/systemd/softether-vpnbridge.service" "${CMAKE_INSTALL_FULL_LIBEXECDIR}/softether/vpnbridge/vpnbridge")
+
+# Copy "vpncmd" directory to /usr/lib(exec)/softether/, install launch script and systemd service
 install(DIRECTORY ${VPNCMD_RUNTIME_OUTPUT_DIRECTORY}
   COMPONENT "vpncmd"
   DESTINATION "${CMAKE_INSTALL_LIBEXECDIR}/softether"
@@ -160,12 +189,7 @@ install(DIRECTORY ${VPNCMD_RUNTIME_OUTPUT_DIRECTORY}
   PERMISSIONS OWNER_READ OWNER_WRITE OWNER_EXECUTE GROUP_READ GROUP_EXECUTE WORLD_READ WORLD_EXECUTE
 )
 
-# Create wrapper scripts in the user's binaries directory, which is usually "/usr/local/bin".
-# This is required because symlinks use the folder they are in as working directory.
-install_wrapper_script("vpnserver" "${CMAKE_INSTALL_FULL_LIBEXECDIR}/softether/vpnserver/vpnserver" ${CMAKE_INSTALL_FULL_BINDIR})
-install_wrapper_script("vpnclient" "${CMAKE_INSTALL_FULL_LIBEXECDIR}/softether/vpnclient/vpnclient" ${CMAKE_INSTALL_FULL_BINDIR})
-install_wrapper_script("vpnbridge" "${CMAKE_INSTALL_FULL_LIBEXECDIR}/softether/vpnbridge/vpnbridge" ${CMAKE_INSTALL_FULL_BINDIR})
-install_wrapper_script("vpncmd" "${CMAKE_INSTALL_FULL_LIBEXECDIR}/softether/vpncmd/vpncmd" ${CMAKE_INSTALL_FULL_BINDIR})
+install_wrapper_script("vpncmd" "${CMAKE_INSTALL_FULL_LIBEXECDIR}/softether/vpncmd/vpncmd")
 
 # Print message after installing the targets
 install(CODE "message(\"\n----------------------------------------------------------------------------------------------------------------------------\")")

--- a/systemd/softether-vpnbridge.service
+++ b/systemd/softether-vpnbridge.service
@@ -1,12 +1,13 @@
 [Unit]
 Description=SoftEther VPN Bridge
 After=network.target auditd.service
-ConditionPathExists=!/opt/vpnbridge/do_not_run
+ConditionPathExists=![DIRECTORY]/do_not_run
 
 [Service]
 Type=forking
-ExecStart=/opt/vpnbridge/vpnbridge start
-ExecStop=/opt/vpnbridge/vpnbridge stop
+EnvironmentFile=-[DIRECTORY]
+ExecStart=[BINARY] start
+ExecStop=[BINARY] stop
 KillMode=process
 Restart=on-failure
 
@@ -15,9 +16,8 @@ PrivateTmp=yes
 ProtectHome=yes
 ProtectSystem=full
 ReadOnlyDirectories=/
-ReadWriteDirectories=-/opt/vpnbridge
+ReadWriteDirectories=-[DIRECTORY]
 CapabilityBoundingSet=CAP_NET_ADMIN CAP_NET_BIND_SERVICE CAP_NET_BROADCAST CAP_NET_RAW CAP_SYS_NICE CAP_SYS_ADMIN CAP_SETUID
 
 [Install]
 WantedBy=multi-user.target
-

--- a/systemd/softether-vpnclient.service
+++ b/systemd/softether-vpnclient.service
@@ -1,13 +1,13 @@
 [Unit]
 Description=SoftEther VPN Client
 After=network.target auditd.service
-ConditionPathExists=!/opt/vpnclient/do_not_run
+ConditionPathExists=![DIRECTORY]/do_not_run
 
 [Service]
 Type=forking
-EnvironmentFile=-/opt/vpnclient
-ExecStart=/opt/vpnclient/vpnclient start
-ExecStop=/opt/vpnclient/vpnclient stop
+EnvironmentFile=-[DIRECTORY]
+ExecStart=[BINARY] start
+ExecStop=[BINARY] stop
 KillMode=process
 Restart=on-failure
 
@@ -16,10 +16,8 @@ PrivateTmp=yes
 ProtectHome=yes
 ProtectSystem=full
 ReadOnlyDirectories=/
-ReadWriteDirectories=-/opt/vpnclient
+ReadWriteDirectories=-[DIRECTORY]
 CapabilityBoundingSet=CAP_NET_ADMIN CAP_NET_BIND_SERVICE CAP_NET_BROADCAST CAP_NET_RAW CAP_SYS_NICE CAP_SYS_ADMIN CAP_SETUID
 
 [Install]
 WantedBy=multi-user.target
-
-

--- a/systemd/softether-vpnserver.service
+++ b/systemd/softether-vpnserver.service
@@ -1,13 +1,13 @@
 [Unit]
 Description=SoftEther VPN Server
 After=network.target auditd.service
-ConditionPathExists=!/opt/vpnserver/do_not_run
+ConditionPathExists=![DIRECTORY]/do_not_run
 
 [Service]
 Type=forking
-EnvironmentFile=-/opt/vpnserver
-ExecStart=/opt/vpnserver/vpnserver start
-ExecStop=/opt/vpnserver/vpnserver stop
+EnvironmentFile=-[DIRECTORY]
+ExecStart=[BINARY] start
+ExecStop=[BINARY] stop
 KillMode=process
 Restart=on-failure
 
@@ -16,10 +16,8 @@ PrivateTmp=yes
 ProtectHome=yes
 ProtectSystem=full
 ReadOnlyDirectories=/
-ReadWriteDirectories=-/opt/vpnserver
+ReadWriteDirectories=-[DIRECTORY]
 CapabilityBoundingSet=CAP_NET_ADMIN CAP_NET_BIND_SERVICE CAP_NET_BROADCAST CAP_NET_RAW CAP_SYS_NICE CAP_SYS_ADMIN CAP_SETUID
 
 [Install]
 WantedBy=multi-user.target
-
-

